File: rest-service/src/main/java/mb/lib/report/db/SelectReportRow.java
package mb.lib.report.db;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.Optional;

import io.vulpine.lib.query.util.basic.BasicPreparedReadQuery;
import mb.lib.report.model.ReportRow;
import mb.lib.model.HashID;

public class SelectReportRow
{
  private static final String Query =
    """
    SELECT
      *
    FROM
      userlogins5.multiblast_fmt_jobs
    WHERE
      report_digest = ?
    """;

  private final Connection con;
  private final HashID reportID;

  public SelectReportRow(Connection con, HashID reportID) {
    this.con      = con;
    this.reportID = reportID;
  }

  public Optional<ReportRow> run() throws Exception {
    return new BasicPreparedReadQuery<>(Query, con, this::parse, this::prepare).execute().getValue();
  }

  private Optional<ReportRow> parse(ResultSet rs) throws Exception {
    if (!rs.next())
      return Optional.empty();

    return Optional.of(Util.rs2Row(rs));
  }

  private void prepare(PreparedStatement ps) throws Exception {
    ps.setBytes(1, reportID.bytes());
  }
}

File: rest-service/src/main/kotlin/mb/lib/report/db/SelectReportRow.kt
package mb.lib.report.db

import io.vulpine.lib.query.util.basic.BasicPreparedReadQuery
import org.veupathdb.lib.hash_id.HashID
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.util.*

data class SelectReportRow(val con: Connection, val reportID: HashID) {
  companion object {
    private val Query =
    """
    SELECT
      *
    FROM
      userlogins5.multiblast_fmt_jobs
    WHERE
      report_digest = ?
    """
  }

  fun run() = BasicPreparedReadQuery(Query, con, this::parse, this::prepare).execute().value!!

  private fun parse(rs: ResultSet) = if (!rs.next()) Optional.empty() else Optional.of(rs2Row(rs))

  private fun prepare(ps: PreparedStatement) = ps.setBytes(1, reportID.bytes)
}
